Don't crash Bazel when a source directory crosses package boundaries

fmeum opened the pull request:
Also add the target for the test that was missed in the merge of https://github.com/bazelbuild/bazel/commit/ad381cb56804d0d0d1dc95ef7bd4da1e4a3a5c90.

Work towards #25834

fmeum commented:
Stacked on #25859

tjgq reviewed (changes requested):
LGTM, please rebase

tjgq commented on src/test/java/com/google/devtools/build/lib/skyframe/SourceDirectoryIntegrationTest.java:
```diff
@@ -228,24 +237,12 @@ public void danglingSymlinkModified_invalidatesAction() throws Exception {
             + " foo/dir/subdir");
   }
 
-  @Test
-  public void infiniteSymlinkExpansion_fails() throws Exception {
-    Path dir = sourceDir.getRelative("subdir/nested2");
-    dir.delete();
-    dir.createSymbolicLink(PathFragment.create(".."));
-    assertThrows(BuildFailedException.class, () -> buildTarget("//foo"));
-    assertContainsEvent("infinite symlink expansion detected");
-    assertContainsEvent("foo/dir/subdir/nested2");
-  }
```
Why was this test deleted?

fmeum replied:
Sorry, I'm still fixing merge conflicts. Is it possible that the test doesn't have a BUILD file entry on master? Will ping you when I've figured this out.

fmeum commented:
@tjgq The commit https://github.com/bazelbuild/bazel/commit/ad381cb56804d0d0d1dc95ef7bd4da1e4a3a5c90 doesn't include the test target for the test I added, but the PR did. I added it here.
